soroban-sdk/src/lib.rs

```diff
@@ -129,7 +129,30 @@ pub use soroban_sdk_macros::symbol;
 macro_rules! panic_error {
     ($env:expr, $error:expr) => {{
         $env.panic_error($error);
-        unreachable!()
+        unreachable!();
+    }};
+}
+
+/// Assert a condition and panic with the given error if it is false.
+///
+/// The first argument in the list must be a reference to an [Env].
+///
+/// The second argument is an expression that if resolves to `false` will cause
+/// a panic with the error in the third argument.
+///
+/// The third argument is an error value. The error value will be given to any
+/// calling contract.
+///
+/// Equivalent to `assert!`, but with an error value instead of a string. The
+/// error value will be given to any calling contract.
+///
+/// See [`contracterror`] for how to define an error type.
+#[macro_export]
+macro_rules! assert_error {
```

Comment:
I chose the name `assert_error` because it parallels `panic_error` (which parallels `panic_any`). I don't love the name `assert_error`, because it reads more like we're asserting that an error should occur, rather than creating an assertion that will cause an error if false. I've gone with consistency, but open to suggestions.

Reply:
I find it a little misleading. You are not asserting that there is an error, rather the opposite.  For near's sdk, they went with `require`, which I would advocate for.

Also I can't wait until macro's can be bound to types! `env.require!(...)` reads better to me.

Reply:
> Also I can't wait until macro's can be bound to types! `env.require!(...)` reads better to me.

Is this in development? Do you have a link to the tracking PR / RFC? I can't find it.

Reply:
It's called [postfix macros](https://github.com/rust-lang/rfcs/pull/2442)  [polyfill here](https://github.com/est31/postfix-macros).

Reply:
@graydon suggested `assert_with_error` or `assert_or_error`.

On it's own I think I prefer `assert_or_error` over require because it retains the assert language.

If we wanted panic_error to also follow, then the `_with_error` variant would probably fit better, i.e. `assert_with_error` and `panic_with_error`.

Thoughts? cc @willemneal @graydon @paulbellamy

Reply:
`*_with_error` makes sense to me.

Reply:
Follow up: https://github.com/stellar/rs-soroban-sdk/pull/752